clippy_lints/src/default_box_assignments.rs

```diff
@@ -0,0 +1,131 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::sugg::Sugg;
+use clippy_utils::ty::implements_trait;
+use clippy_utils::{is_default_equivalent_call, local_is_initialized, path_def_id, path_to_local};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects assignments of `Default::default()` or `Box::new(value)`
+    /// to a place of type `Box<T>`.
+    ///
+    /// ### Why is this bad?
+    /// This incurs an extra heap allocation compared to assigning the boxed
+    /// storage.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// b = Default::default();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let mut b = Box::new(1u32);
+    /// *b = Default::default();
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub DEFAULT_BOX_ASSIGNMENTS,
+    perf,
+    "assigning `Default::default()` to `Box<T>` is inefficient"
+}
+declare_lint_pass!(DefaultBoxAssignments => [DEFAULT_BOX_ASSIGNMENTS]);
+
+impl LateLintPass<'_> for DefaultBoxAssignments {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        if let ExprKind::Assign(lhs, rhs, _) = &expr.kind {
+            let lhs_ty = cx.typeck_results().expr_ty(lhs);
+
+            // No diagnostic for late-initialized locals
+            if let Some(local) = path_to_local(lhs)
+                && !local_is_initialized(cx, local)
+            {
+                return;
+            }
+
+            let Some(inner_ty) = get_box_inner_type(cx, lhs_ty) else {
+                return;
+            };
+
+            let Some(default_trait_id) = cx.tcx.get_diagnostic_item(sym::Default) else {
+                return;
+            };
+
+            if implements_trait(cx, inner_ty, default_trait_id, &[])
+                && is_default_call(cx, rhs)
+                && !rhs.span.from_expansion()
```

Comment:
`rhs.span_from_expansion()` should be excluded from the start, as the user might not be able to remove the `Box::default()` or `Box::new()` if it comes from a macro. Same for the `lhs` side. In fact, neither `expr` nor `lhs` nor `rhs` should come from a macro (and tests should be added to check this).